### ISSUE TITLE
💬 RFC:  Backstage Enhancement Proposals (BEP) process update

### ISSUE THREAD

**Issue by Rugvip** [backstage/backstage#27267]
### 🔖 Need

[BEPs](https://github.com/backstage/backstage/tree/master/beps) have been a part of our process in the project for a little while now. I find it to be fairly successful in discussing work to be picked up and shaping the needs/problems that we're trying to solve. 

On the other hand I don't find it to work that well when it comes to iterating on existing proposals and keeping them in sync with experimental implementations. There's in particular a bit of a disconnect in expectations when it comes to what it means to have a proposal merged as provisional vs moving it to implementable.

### 🎉 Proposal

Merge BEPs way later. We can still discuss the feasability of proposals very early on in PRs, but once we merge a BEP I feel we loose much over the overview and discussion. My proposal is that we effectively keep BEP PRs open until we would've hit the `implementable` state, removing the `provisional` state altogether. Merging the BEP then instead signals that we have settled the design and are ready to implement it, while rejecting the PR is a signal that the proposal is not yet ready.

### 〽️ Alternatives

We could keep the existing setup, but move BEPs to a separate repo. That might help out isolating discussions better.

### ❌ Risks

Nothing major here that I can see, just looking for comments 😁 

### 👀 Have you spent some time to check if this RFC has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

**Comment by aramissennyeydd** [comment]
Thanks for opening this! I'm likely a problem child in this area 😁. I think this solves a small part of the frustrating experience of being a BEP owner, but I'd love for us to solve more. As a BEP owner, I'm looking to understand,

a) Is my idea worth pursuing and is there interest/resources from the maintainer team to review my proposal and following PRs?
b) What does lack of feedback on my proposal or changes mean for my BEP?
c) Is my design solid and in line with Backstage standards? How do I resolve issues with my design?
d) Am I allowed to start opening PRs against Backstage proper with changes from my design?

I believe this was the original idea with merging BEPs' initial drafts quickly. Merging a proposal was intended to show that there was interest to pursue it and time from the maintainer team to advise/review. In the proposed version, this gets a bit muddier. This also makes it harder to find BEPs that were interesting but we didn't have resources to pursue (`deferred` status).

I'd propose explicitly using SIG time for BEP review. This idea came up recently at the Docs SIG. Most SIGs already cordon off time for PR/RFC reviews, what if we added earmarked time for BEP reveiws? The goals of this section would be to
1. Reduce the amount of upfront risk an author is taking on writing a BEP - SIGs would allow time for BEP pitches, maintainers could ask questions and help steer the author in the right direction.
2. Connects reviewers to authors.
3. Keep progress moving and BEPs unblocked - BEP authors would commit to a x minute slot for progress updates, design questions and experiment findings. This would offer realtime feedback on stalled or at-risk BEPs that may need some more love.

**Comment by Rugvip** [comment]
@aramissennyeydd thank you!

Going bit by bit:

> a) Is my idea worth pursuing and is there interest/resources from the maintainer team to review my proposal and following PRs?

This should be immediately communicated by maintainers as soon as a BEP/RFC is opened, I think this is on maintainers to fix if it's broken.

> b) What does lack of feedback on my proposal or changes mean for my BEP?

Often this is gonna be because the proposal doesn't solve a problem very high priority, but still is valuable. It lies in the non-sweetspot of the backlog where it's too good to reject but too much effort to take on straight away. This comes back to a).

> c) Is my design solid and in line with Backstage standards? How do I resolve issues with my design?

That'll be evident in feedback, and is usually fairly fast I'd say. The tricky area is when is when a BEP proposes changed or new standards, that takes a lot of time to figure out.

> d) Am I allowed to start opening PRs against Backstage proper with changes from my design?

Not until the BEP is marked as `implementable`, unless the changes are shipped as experimental. I think the lines are a bit blurry when it comes to the experimental part - often experiments are required to settle the design, but at the same they tend to stick around. I think this is a problem that's more related to how we do experiments rather than the BEP process.

> This also makes it harder to find BEPs that were interesting but we didn't have resources to pursue

I think browsing closed PRs with the `bep` label can be good enough here tbh.

> I'd propose explicitly using SIG time for BEP review. 

Love it, I think BEP reviews should get priority above other PR reviews in SIGs. That also includes discussing new BEPs and whether it'd be worth pursuing them. We do want to make sure that this also helps those that don't want/can't join SIG meetings, so we likely want to make this a general BEP review rather than only on request. I assume it's your intention and that it's the case that this change would help solve for a), b), and c) from the above?

---

Coming back to this I'm fairly bullish still on removing the `provisional` state of BEPs though. The initial BEP PR I think is a very good place for discussion, but as soon as it's merged that goes away. The BEP discussion issue and followup PRs aren't at all as effective at driving discussion, and it's easy for things to fizzle out. Merging straight into `implementable` also means that we can assign ownership of the implementation at that point, and potential deadlines. With that I think we'd also not create a discussion issue, but instead a meta issue for each BEP with a task list that is tracking the path to completing the implementation.

**Comment by aramissennyeydd** [comment]
> I think this is a problem that's more related to how we do experiments rather than the BEP process.

100%, this may get better over time as there'll be more examples of BEPs done well.

> We do want to make sure that this also helps those that don't want/can't join SIG meetings, so we likely want to make this a general BEP review rather than only on request. I assume it's your intention and that it's the case that this change would help solve for a), b), and c) from the above?

Yup 😁! For those that can't join the meeting, we can fallback to just PR discussion/Discord, but I want the SIG to be a discussion place for these topics, not just a synchronous meeting I can go to to get unblocked on slow PR feedback or deadlock.

> Coming back to this I'm fairly bullish still on removing the provisional state of BEPs though.

I think we should do both! It would help with more direct feedback on when the design is finished and implementation is ready to start as I've struggled with this myself.
